builds/jenkins-delete-resources/Jenkinsfile

```diff
@@ -0,0 +1,554 @@
+  for (account in finalAccounts) {
+   credentialId = account['CREDENTIAL_ID'] 
+   def accountID = account['ACCOUNTID'];
+   def iamRoles = account["IAM"]
+  def cloudFront = account["CLOUDFRONT"]
+   isPrimaryAccount = account['PRIMARY']
+    withCredentials([
+      [$class: 'AmazonWebServicesCredentialsBinding', credentialsId: "${credentialId}"]
+    ]) {
+       
+        // Get all Regions
+        def allregions = account["REGIONS"]
+        for (awsregion in allregions) {
+        echo "processing  Region ${awsregion['REGION']}"
+        def regionAws = awsregion['REGION']
+        //Check if the Region is Primary
+        def primaryRegion = awsregion['PRIMARY']
+        if( isPrimaryAccount && primaryRegion){
+          echo "Primary Account Primary cannot be deleted from the Job"
+        }else{
+          //Get Region based apiGateway and s3Buckets
+          def apiGateway = awsregion["API_GATEWAY"]
+          def s3Buckets = awsregion["S3"]
+          processApiGateway(apiGateway, regionAws)
+          processS3Buckets(s3Buckets)
+          deletePrimaryAccountPolicy(accountID, finalAccounts);
+          deleteUserServices(regionAws);
+          deleteStackCloudFrontDists(regionAws)
+        }
+      }
+      if ( !isPrimaryAccount ) {
+          processIamRoles(iamRoles);
+          processCloudFront(cloudFront);
+      }
+    }
+  }
+ } else {
+  echo "No Accounts available for Deletion"
+ }
+}
+
+def processIamRoles(iamRoles) {
+ def roleId = ""
+ // Get the Iam roles and delete it
+ iamRoles.each {
+  k,
+  v ->
+  if (v.contains('/')) {
+   roleId = v.split("/");
+  }
+
+  echo "Processing Iam Role of ${k}"
+  if (roleId[1] != "") {
+   deleteIam(roleId[1])
+  } else {
+   echo "Not a Valid Role Name"
+  }
+ }
+}
+
+def processCloudFront(cloudfront) {
+ def distribution_id = ""
+
+ // Get the Iam roles and delete it
+ cloudfront.each {
+  k,
+  v ->
+   deletecloudFront(v)
+ }
+}
+
+def processApiGateway(apigateway, regionAws) {
+ // Get the Iam roles and delete it
+ apigateway.each {
+  k,
+  v ->
+  echo "Processing API Gateway  of ${k}"
+  deleteAPIGateway(v, regionAws)
+ }
+}
+
+def processS3Buckets(s3Buckets) {
+ // Get the s3Buckets and delete it
+ s3Buckets.each {
+  k,
+  v ->
+  echo "Processing s3Buckets of ${k}"
+  deletes3Bucket(v)
+ }
+}
+/**
+  Function for Deleting IAM Roles
+  Accepts credentialId and Role Name
+  Check if Role exists and delete it
+**/
+def deleteIam(roleName) {
+ try {
+  def status = true;
+  try {
+   sh "aws iam get-role --role-name ${roleName}"
+  } catch (ex) {
+   echo "Iam $roleName is empty"
+   status = false
+  }
+  // If Iam exists delete it
+  if (status) {
+   //check and delete Custom Policies attached 
+   deleteInlinePolicies(roleName)
+   //check and delete if policies attached
+   detachIamPolicies(roleName)
+   //Delete the role
+   deleteIamRole(roleName)
+  } else {
+   echo "Iam named ${roleName} not found"
+  }
+ } catch (ex) {
+  echo "${ex.getMessage()}"
```

Comment:
Question - Are these exceptions being caught and events being sent (as applicable/appropriate)?

Reply:
Events are not implemented in this Job, I am just logging the exception.